### ISSUE TITLE
【AIコース】大課題 ボストン住宅価格 線形回帰

### ISSUE THREAD

**Issue by hayyoshi** [hayyoshi/DIC#1]
### 線形回帰とは何か
# 入力ベクトル(特徴量)と対応する教師ベクトル(答え)の関係を表現する関数y(x)を推測する。
# データと用意した関数のずれ(誤差)を最小にする傾きと切片を求める。
# ずれは、誤差関数により測定する事が可能。
# 予測は用意した関数に従った連続値となる。一方、分類は用意した項目にカテゴライズする。

### データを取得する
from sklearn.datasets import load_boston
### 取得データをDataFrameにする
boston = load_boston()
import pandas as pd
X = pd.DataFrame(boston.data,columns=boston.feature_names)
y = pd.DataFrame(boston.target)
### 説明変数を'LSTAT'のみにする
x = X['LSTAT']
### 単回帰と重回帰についての違いを記述せよ
# 単回帰は特徴量(または説明変数と呼ぶ)が１つ、重回帰は複数という違いがある。
### テストデータに分割する
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(x,y,test_size=0.2)
### 学習
from sklearn.linear_model import LinearRegression
lin_1d = LinearRegression()
lin_1d.fit(X_train[:,None],y_train)
### 決定係数
score_1d = lin_1d.score(X_test[:,None],y_test)
print("一次式における'LSTAT'の住宅価格への決定係数は%.2f"%(score_1d))
### 決定係数とは何か記述せよ
# モデルの汎化能力を表す。最大1を取り、1に近いほど性能が良い事を表す。
# 本数値から、選択した特徴量(説明変数)が目的変数をどれだけ説明出来ているかを判断出来る。
# つまり、もっとも説明変数が、目的変数を説明できる場合、決定係数は1になる。
# 決定係数は以下の式から求める事が出来る。
# R**2 = 1-最小2乗誤差を分散で割った値

### 決定係数をいかなる場合も信じても良いか記述せよ
# 特徴量(説明変数)を増やして決定係数が良くなったとしても、
# 分析を行う目的や、検証したい仮説とは一切関係ない変数を無秩序に分析に加えてしまうと、
# そのモデルは解釈が難しくなり当初の目的を果たせない。
# よって、決定係数は絶対の指標ではない。

### 2,3,4次式の回帰
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

lin_2d = LinearRegression()
lin_3d = LinearRegression()
lin_4d = LinearRegression()

degree_2 = PolynomialFeatures(degree=2)
degree_3 = PolynomialFeatures(degree=3)
degree_4 = PolynomialFeatures(degree=4)

x_train_2 = degree_2.fit_transform(X_train[:,None])
x_train_3 = degree_3.fit_transform(X_train[:,None])
x_train_4 = degree_4.fit_transform(X_train[:,None])

x_test_2 = degree_2.fit_transform(X_test[:,None])
x_test_3 = degree_3.fit_transform(X_test[:,None])
x_test_4 = degree_4.fit_transform(X_test[:,None])

lin_2d.fit(x_train_2,y_train)
lin_3d.fit(x_train_3,y_train)
lin_4d.fit(x_train_4,y_train)

score_2d = lin_2d.score(x_test_2,y_test)
score_3d = lin_3d.score(x_test_3,y_test)
score_4d = lin_4d.score(x_test_4,y_test)

print("二次式における'LSTAT'の住宅価格への決定係数は%.2f"%(score_2d))
print("三次式における'LSTAT'の住宅価格への決定係数は%.2f"%(score_3d))
print("四次式における'LSTAT'の住宅価格への決定係数は%.2f"%(score_4d))

### 次数が大きくなるとどうなるか記述せよ
# 一次式：ax + b 二次式：ax*2 + bx + c 三次式：ax*3 + bx*2 + cx + dとなり、
# 次数を増やすメリットとして、実際のデータを説明できるような曲線を描ける事が挙げられる。
# 一方、次数を増やした事で過学習が発生するデメリットが挙げられる。

### 重回帰
Multiple_X=pd.concat([X,y],axis=1)
Multiple_X.corr()
type(X.loc[:,['RM','LSTAT','ZN']].as_matrix())
x2 = X.loc[:,['RM','LSTAT','ZN']]

from sklearn.model_selection import train_test_split
X2_train,X2_test,y2_train,y2_test = train_test_split(x2,y,test_size=0.2)

lin_2d_m = LinearRegression()
lin_3d_m = LinearRegression()
lin_4d_m = LinearRegression()

degree_2_m = PolynomialFeatures(degree=2)
degree_3_m = PolynomialFeatures(degree=3)
degree_4_m = PolynomialFeatures(degree=4)

x2_train_2 = degree_2_m.fit_transform(X2_train.as_matrix())
x2_train_3 = degree_3_m.fit_transform(X2_train.as_matrix())
x2_train_4 = degree_4_m.fit_transform(X2_train.as_matrix())

x2_test_2 = degree_2.fit_transform(X2_test.as_matrix())
x2_test_3 = degree_3.fit_transform(X2_test.as_matrix())
x2_test_4 = degree_4.fit_transform(X2_test.as_matrix())

lin_2d_m.fit(x2_train_2,y2_train)
lin_3d_m.fit(x2_train_3,y2_train)
lin_4d_m.fit(x2_train_4,y2_train)

score_2d_x2 = lin_2d_m.score(x2_test_2,y2_test)
score_3d_x2 = lin_3d_m.score(x2_test_3,y2_test)
score_4d_x2 = lin_4d_m.score(x2_test_4,y2_test)

print("二次式における'RM','LSTAT','ZN'の住宅価格への決定係数は%.2f"%(score_2d_x2))
print("三次式における'RM','LSTAT','ZN'の住宅価格への決定係数は%.2f"%(score_3d_x2))
print("四次式における'RM','LSTAT','ZN'の住宅価格への決定係数は%.2f"%(score_4d_x2))

### 重回帰について記述せよ
# 説明変数を増やす事で、目的変数をより詳細に説明出来るようになる為、予測の精度が上がるメリットがある。 
# 一方、説明変数を増やした際に、目的変数と関係のない説明変数だった場合、予測の精度が悪くなるデメリットがある。
